panic: runtime error: invalid memory address or nil pointer dereference

When creating a new cluster, performing some edits to the ig and then running an update we get the following:

```
ubuntu@ip-10-157-38-30:~$ work/bin/kops update cluster --state s3://s3-bucket --name kube-test.our-domain.com --yes
I0120 11:07:49.819437    1689 populate_cluster_spec.go:235] Defaulting DNS zone to: ZGQJ6GJOGZN3
I0120 11:07:52.015846    1689 executor.go:91] Tasks: 0 done / 111 total; 35 can run
I0120 11:07:53.174897    1689 vfs_castore.go:422] Issuing new certificate: "master"
I0120 11:07:53.264387    1689 vfs_castore.go:422] Issuing new certificate: "kubelet"
I0120 11:07:53.868113    1689 vfs_castore.go:422] Issuing new certificate: "kubecfg"
I0120 11:07:55.928938    1689 executor.go:91] Tasks: 35 done / 111 total; 23 can run
I0120 11:07:57.329473    1689 executor.go:91] Tasks: 58 done / 111 total; 36 can run
W0120 11:07:59.647164    1689 executor.go:109] error running task "LaunchConfiguration/master-us-west-2a.masters.kube-test.our-domain.com" (9m57s remaining to succeed): IAM instance profile not yet created/propagated (original error: You are not authorized to perform this operation.)
W0120 11:07:59.647288    1689 executor.go:109] error running task "LaunchConfiguration/master-us-west-2c.masters.kube-test.our-domain.com" (9m57s remaining to succeed): IAM instance profile not yet created/propagated (original error: You are not authorized to perform this operation.)
W0120 11:07:59.647349    1689 executor.go:109] error running task "LaunchConfiguration/nodes.kube-test.our-domain.com" (9m57s remaining to succeed): IAM instance profile not yet created/propagated (original error: You are not authorized to perform this operation.)
W0120 11:07:59.647403    1689 executor.go:109] error running task "LaunchConfiguration/bastions.kube-test.our-domain.com" (9m57s remaining to succeed): IAM instance profile not yet created/propagated (original error: You are not authorized to perform this operation.)
W0120 11:07:59.647468    1689 executor.go:109] error running task "LaunchConfiguration/master-us-west-2b.masters.kube-test.our-domain.com" (9m57s remaining to succeed): IAM instance profile not yet created/propagated (original error: You are not authorized to perform this operation.)
I0120 11:07:59.647538    1689 executor.go:91] Tasks: 89 done / 111 total; 10 can run
I0120 11:08:00.859445    1689 logging_retryer.go:50] Retryable error 400 (400 Bad Request) from autoscaling/CreateLaunchConfiguration - will retry after delay of 676ms
I0120 11:08:00.867789    1689 logging_retryer.go:50] Retryable error 400 (400 Bad Request) from autoscaling/CreateLaunchConfiguration - will retry after delay of 589ms
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa13f84]

goroutine 1080 [running]:
panic(0x1a8f5c0, 0xc4200120b0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
k8s.io/kops/upup/pkg/fi/cloudup/awstasks.(*NatGateway).RenderAWS(0xc420c36ac0, 0xc42012cb00, 0x0, 0xc420c36ac0, 0xc421035e80, 0x0, 0x0)
	/home/ubuntu/work/src/k8s.io/kops/upup/pkg/fi/cloudup/awstasks/natgateway.go:316 +0x144
reflect.Value.call(0x1c7a300, 0xc420c36ac0, 0x1213, 0x1d37f1d, 0x4, 0xc420d16900, 0x4, 0x4, 0xc420ae9b78, 0x13, ...)
	/usr/local/go/src/reflect/value.go:434 +0x5c8
reflect.Value.Call(0x1c7a300, 0xc420c36ac0, 0x1213, 0xc420d16900, 0x4, 0x4, 0xc420c36ac0, 0x1213, 0xc420d168a0)
	/usr/local/go/src/reflect/value.go:302 +0xa4
k8s.io/kops/upup/pkg/fi.(*Context).Render(0xc4204a38f0, 0x2c0c8c0, 0x0, 0x2c0c8c0, 0xc420c36ac0, 0x2c0c8c0, 0xc421035e80, 0x0, 0xc41ffb7470)
	/home/ubuntu/work/src/k8s.io/kops/upup/pkg/fi/context.go:154 +0xd0c
k8s.io/kops/upup/pkg/fi.DefaultDeltaRunMethod(0x2c0c8c0, 0xc420c36ac0, 0xc4204a38f0, 0xc4209171e0, 0xc4201d0c48)
	/home/ubuntu/work/src/k8s.io/kops/upup/pkg/fi/default_methods.go:56 +0x568
k8s.io/kops/upup/pkg/fi/cloudup/awstasks.(*NatGateway).Run(0xc420c36ac0, 0xc4204a38f0, 0x16, 0xc420aee758)
	/home/ubuntu/work/src/k8s.io/kops/upup/pkg/fi/cloudup/awstasks/natgateway.go:250 +0x41
k8s.io/kops/upup/pkg/fi.(*executor).forkJoin.func1(0xc420b36aa0, 0xa, 0xa, 0xc420916ce0, 0xc420028628, 0xc4201d0c40, 0x0)
	/home/ubuntu/work/src/k8s.io/kops/upup/pkg/fi/executor.go:158 +0x1df
created by k8s.io/kops/upup/pkg/fi.(*executor).forkJoin
	/home/ubuntu/work/src/k8s.io/kops/upup/pkg/fi/executor.go:159 +0x132
```

The only edits we're making are to the cloudLabels on some of the ig's and adding the connection timeout setting for the bastions ELB.

Version:

```
Version 1.5.0-alpha3 (git-b749173)
```

--- a/upup/pkg/fi/cloudup/awstasks/natgateway.go
+++ b/upup/pkg/fi/cloudup/awstasks/natgateway.go
@@ -43,9 +43,10 @@ type NatGateway struct {
 	AssociatedRouteTable *RouteTable
 }
 
-var _ fi.CompareWithID = &NatGateway{} // Validate the IDs
+var _ fi.CompareWithID = &NatGateway{}
 
 func (e *NatGateway) CompareWithID() *string {
+	// Match by ID (NAT Gateways don't have tags, so they don't have a name in EC2)
 	return e.ID
 }
 
@@ -55,7 +56,7 @@ func (e *NatGateway) Find(c *fi.Context) (*NatGateway, error) {
 	var ngw *ec2.NatGateway
 	actual := &NatGateway{}
 
-	if e.ID != nil && *e.ID != "" {
+	if fi.StringValue(e.ID) != "" {
 		// We have an existing NGW, lets look up the EIP
 		var ngwIds []*string
 		ngwIds = append(ngwIds, e.ID)
@@ -71,23 +72,18 @@ func (e *NatGateway) Find(c *fi.Context) (*NatGateway, error) {
 		}
 
 		if len(response.NatGateways) != 1 {
-			return nil, fmt.Errorf("found %b NAT Gateways, expected 1", len(response.NatGateways))
+			return nil, fmt.Errorf("found %d Nat Gateways, expected 1", len(response.NatGateways))
 		}
-		if len(response.NatGateways) == 1 {
-			ngw = response.NatGateways[0]
-		}
+		ngw = response.NatGateways[0]
 
-		if len(response.NatGateways[0].NatGatewayAddresses) != 1 {
-			return nil, fmt.Errorf("found %b EIP Addresses for 1 NAT Gateway, expected 1", len(response.NatGateways))
+		if len(ngw.NatGatewayAddresses) != 1 {
+			return nil, fmt.Errorf("found %d EIP Addresses for 1 NATGateway, expected 1", len(ngw.NatGatewayAddresses))
 		}
-		if len(response.NatGateways[0].NatGatewayAddresses) == 1 {
-
-			actual.ElasticIP = &ElasticIP{ID: response.NatGateways[0].NatGatewayAddresses[0].AllocationId}
-
-		}
+		actual.ElasticIP = &ElasticIP{ID: ngw.NatGatewayAddresses[0].AllocationId}
 	} else {
 		// This is the normal/default path
-		ngw, err := e.findNatGateway(c)
+		var err error
+		ngw, err = e.findNatGateway(c)
 		if err != nil {
 			return nil, err
 		}
@@ -229,6 +225,9 @@ func (s *NatGateway) CheckChanges(a, e, changes *NatGateway) error {
 		if e.Subnet == nil {
 			return fi.RequiredField("Subnet")
 		}
+		if e.AssociatedRouteTable == nil {
+			return fi.RequiredField("AssociatedRouteTable")
+		}
 	}
 
 	// Delta
@@ -300,7 +299,7 @@ func (_ *NatGateway) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *NatGateway)
 		return fmt.Errorf("Subnet ID not set")
 	}
 
-	// TODO: Obsolete - we can get from the route table instead
+	// TODO: AssociatedNatgateway tag is obsolete - we can get from the route table instead
 	tags := make(map[string]string)
 	tags["AssociatedNatgateway"] = *id
 	err := t.AddAWSTags(*e.Subnet.ID, tags)
@@ -313,9 +312,12 @@ func (_ *NatGateway) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *NatGateway)
 	// This is better than just a tag that's shared because this lets us create a whitelist of these NGWs
 	// without doing a bunch more work in `kutil/delete_cluster.go`
 
-	if *e.Shared == true {
-		glog.V(2).Infof("tagging route table %s to track shared NGW", *e.AssociatedRouteTable.ID)
-		err = t.AddAWSTags(*e.AssociatedRouteTable.ID, tags)
+	if fi.BoolValue(e.Shared) {
+		if e.AssociatedRouteTable == nil {
+			return fmt.Errorf("AssociatedRouteTable not provided")
+		}
+		glog.V(2).Infof("tagging route table %s to track shared NGW", fi.StringValue(e.AssociatedRouteTable.ID))
+		err = t.AddAWSTags(fi.StringValue(e.AssociatedRouteTable.ID), tags)
 	}
 
 	return nil

--- a/upup/pkg/fi/cloudup/awsup/aws_cloud.go
+++ b/upup/pkg/fi/cloudup/awsup/aws_cloud.go
@@ -218,6 +218,10 @@ func (c *awsCloudImplementation) GetTags(resourceID string) (map[string]string, 
 }
 
 func getTags(c AWSCloud, resourceId string) (map[string]string, error) {
+	if resourceId == "" {
+		return nil, fmt.Errorf("resourceId not provided to getTags")
+	}
+
 	tags := map[string]string{}
 
 	request := &ec2.DescribeTagsInput{

